OpenGL Tutorials (#1123)
dotnet/Silk.NET: website/docs/opengl/c1/2-hello-quad.md

## Clearing the window
Before we start this, let's take a look at what makes up an OpenGL window. 

A window consists of several **framebuffers**. A framebuffer is essentially a texture that can be drawn to by OpenGL. An OpenGL window will consist of the following framebuffers:

* Color buffer
* Depth stencil buffer

On top of this, a window will often contain at least two of these buffers. This is known as **double-buffering** and is imperitive for VSync to work properly. One buffer is displayed, while another is rendered to. 

This is what makes up the **swapchain**, which, if you've ever looked into Direct3D, you should be familliar with. OpenGL does not use the term, nor does it allow you to manage this yourself, however it is still in the background, and it is helpful to know this.

Now that we know this, we can get on with clearing the window.

Add the following using directive to the top of your file:

```cs
using System.Drawing;
```

Then, in your `OnLoad` method, underneath `GetApi()`, add the following:

```cs
_gl.ClearColor(Color.CornflowerBlue);

> ```

Reviewer: Silk.NET is a 1:1 binding of OpenGL, and as a result the names are mapped 1:1. Please consult the OpenGL documentation for any clarifications on the bindings themselves.

(...and we're certainly not going to rename anything within 2.X as that'd be massively breaking!)

> ```

Reviewer 2: Might be worth suggesting on 3.0 in one of the meetings though.

> ```

Author: Can't imagine adding it into 3.0 is a good idea, as again that would go against the 1:1 bindings, and would certainly annoy plenty of users, myself included.

If some people wanted this type of functionality, perhaps a small library could be written over the GL functions to do this behaviour instead, and the developer could apply to have it as part of SilkCommunity (I'm not sure what the guidelines are there..). Just an idea, I personally won't be developing anything like that since `_gl.ClearColor = Color.CornflowerBlue;` is no real different to `_gl.ClearColor(Color.CornflowerBlue)`, so imo the change is a bit pointless.

---
{
    "TableOfContents": {
        "Name": "1.2 - Hello Quad",
        "Url": "2-hello-quad.html",
        "Metadata": {
            "AuthorGitHub": "ohtrobinson",
            "DateTimeWritten": "02/11/2022 12:00",
            "PreviewImage": ""
        }
    }
}
---

# 1.2 - Hello Quad

> <?# Info "You can view the source code for this tutorial [here.](../sources/1.2-final-result.html) This tutorial builds on the previous tutorial. If you haven't read it, you can do so [here.](1-hello-window.html)" /?>

Reviewer 3: ```suggestion
<?# Info "You can view the source code for this tutorial [here](../sources/1.2-final-result.html). This tutorial builds on the previous tutorial. If you haven't read it, you can do so [here](1-hello-window.html)." /?>
```

Let's draw something on-screen! In this tutorial, you'll learn:

* How to initialize OpenGL.
* How to clear the window.
* How the OpenGL pipeline works.
* What vertex array objects, vertex buffers, and element array buffers are, and how to use them.
* How to create and use shaders.
* How to draw to the screen.

This tutorial will feature a lot of content and new info, so we'll take it slow and with lots of explanations.

## Initializing OpenGL
The first thing we need to do before we can issue OpenGL commands, is to initialize it. Fortunately, Silk.NET windowing does most of the hard part for us, and we only have to get the GL api.

Add the following using directive to the top of your file, you'll need this to use OpenGL:

```cs
using Silk.NET.OpenGL;
```

Then, at the top of your class, add the following:

```cs
private static GL _gl;
```

Finally, in your `OnLoad` method, add the following:

```cs
_gl = _window.CreateOpenGL();
```

> What are we doing here? Silk.NET requires you to keep a **reference** to the OpenGL API. If you've used or seen OpenGL in C, you'll notice that this is different to the way that it is done there. This is done so that you can more easily keep track of multiple contexts. If you don't know what that is, don't worry about it for now, we won't be using it in these tutorials.

Reviewer 3: > This is done so that you can more easily keep track of multiple contexts.

@Perksey Just wanna check with you that this is correct

Reviewer: This is probably the most obvious reason yeah.

<?# Info "If you take a look at the source in the Silk.NET samples repository, you will notice that it uses `_gl = GL.GetApi(_window)`. This is another way to get the GL API, however when using Silk.NET windowing, it's recommended that you use `_window.CreateOpenGL()` instead." /?>

Now, run your application again. If all is good, you should see no change. Awesome! Let's do our first steps in OpenGL: Clearing the window.

## Clearing the window
Before we start this, let's take a look at what makes up an OpenGL window. 

> A window consists of several **framebuffers**. A framebuffer is essentially a texture that can be drawn to by OpenGL. An OpenGL window will consist of the following framebuffers:
> 
> * Color buffer
> * Depth stencil buffer
> 

Reviewer 3: ```suggestion
A window consists of several default **framebuffers**. A framebuffer is a set of buffers that can be rendered to, essentially a texture that can be drawn to by OpenGL. An OpenGL window's default framebuffers will consist of the following buffers:

* Color buffer
* Depth stencil buffer
```

According to [the khronos docs](https://www.khronos.org/opengl/wiki/Framebuffer):
> A Framebuffer is a collection of buffers that can be used as the destination for rendering.

So a framebuffer _contains_ the color, depth and stencil buffers. Still, let's keep the "essentially a texture that can be drawn to" because that's much easier to understand for newbies :)

Author: I agree about changing it to a "set of buffers", that is an oversight on my part. However I dislike calling them "default framebuffers", since, while yes the window has a default framebuffer, I do mention (i think?) that you can't directly interact with the window's framebuffer. And for example in Direct3D you *can* interact with the window's framebuffer, I prefer to say that the window consists of several framebuffers instead of *default* framebuffer. (I could go on with a long rant about the fact that you can't directly interact with the window's framebuffer... just give me the damn framebuffer so I can mess with it...!)

Reviewer 3: Fair enough lol. _Damn you, window framebuffers!_

> On top of this, a window will often contain at least two of these buffers. This is known as **double-buffering** and is imperitive for VSync to work properly. One buffer is displayed, while another is rendered to. 

Reviewer 3: I don't think double-buffering is specifically about VSync. If you're always rendering to the same framebuffer, then what will the screen display? The resulting frame will only exist for a brief period of time between you finishing your draw calls for that frame and you clearing the framebuffer to start the next. You can't display a framebuffer that's currently being worked on!

Reviewer 3: ```suggestion
On top of this, a window will often contain at least two sets of these buffers. This is known as **double-buffering** and is imperitive for VSync to work properly. One buffer is displayed, while another is rendered to. 
```

Author: Yeah you're right, hadn't really thought about it outside of vsync. I'll change this to make it clearer tomorrow.

Author: I've removed the bits about vsync, and replaced it to say "imperative for rendering to work properly."

This is what makes up the **swapchain**, which, if you've ever looked into Direct3D, you should be familliar with. OpenGL does not use the term, nor does it allow you to manage this yourself, however it is still in the background, and it is helpful to know this.

Now that we know this, we can get on with clearing the window.

Add the following using directive to the top of your file:

```cs
using System.Drawing;
```

Then, in your `OnLoad` method, add the following:

```cs
_gl.ClearColor(Color.CornflowerBlue);
```

> You may notice that this function contains various **overloads**, including ones for a Vector4, and 4 floats. We'll be using the overload that takes `System.Drawing.Color`, because it is the easiest to use and understand.

Reviewer 3: ```suggestion
You may notice that this function contains various **overloads**, including ones for a `System.Numerics.Vector4`, and 4 floats. We'll be using the overload that takes `System.Drawing.Color`, because it is the easiest to use and understand.
```

If you run your application now, you'll notice that the window is still black. That's because we've set the clear *color*, but not actually told OpenGL to clear the window.

In your `OnRender` method, add the following:

```cs
_gl.Clear(ClearBufferMask.ColorBufferBit);
```

Run your application again, and you should see a lovely sky blue window!

![Sky blue window](../../../images/opengl/chapter1/cornflower-window.png)

Congrats! You've done your first thing in OpenGL! Didn't work? Check the [source code](../sources/1.2.2-clear-window.html) for this section here.

But - how does OpenGL know to color the window blue? Sure, we've told it we want blue - but we only did that once... How does it know to keep using blue?

Well, this is because OpenGL is a **state machine**.

> ## OpenGL: The state machine

Reviewer 3: I like the smooth transition!

Before we can continue further, you'll have to get used to the concept that OpenGL is a **state machine**. It's really hard to understand OpenGL if you don't understand this, so read closely.

### What is a state machine?
At its core, a state machine holds... state. You set the state of something, and it retains that state until you change it.

This is exactly how OpenGL works. Once you set something, it will remain set until you change it. This counts for everything in OpenGL. Clear color, binding objects, etc etc, everything goes through the state machine. You can *manipulate* the current state, however you have to be wary at all times of what part of the state you are changing. Change the wrong thing and suddenly you program might not work!

So, this explains why clearing the window works. You set the clear color, and it remains *as* the clear color until you change it.

Now that we've got that, let's move on!

## Vertex Array Objects (VAOs)

> A unique feature that modern OpenGL has is what is known as a Vertex Array Object, or VAO. It stores the state of various parameters, such as which buffers you wish to use, and the shader inputs.

Reviewer 3: ```suggestion
A unique feature that modern OpenGL has is what is known as a Vertex Array Object, or VAO. It stores all of the state needed to supply vertex data, such as which buffers you wish to use and how the vertex data is read from said buffers.
```
I think this will be clearer, what do you think?

From [the khronos docs](https://www.khronos.org/opengl/wiki/Vertex_Specification):
> A Vertex Array Object (VAO) is an [OpenGL Object](https://www.khronos.org/opengl/wiki/OpenGL_Object) that stores all of the state needed to supply vertex data

Author: It is clearer, my only concern is that it may confuse some people who don't understand what vertices are, since that hasn't been covered at this part of the tutorial - I'll include something along these lines but I'll try to make it easier to understand for newbies, perhaps something like `It stores all the necessary information required to draw an object to the screen, such as the vertex data (which we'll get into in a minute), and also information on how to process this data.`. How does that sound?

Reviewer 3: Not sure about the part on "how to process this data". All a VAO stores is the "how to read the vertex data". How to process is part of the shader program, which VAOs know nothing about.

Author: whoops, yes, this has now been fixed.

This is a required feature of modern OpenGL. You **must** have a VAO bound, otherwise your application won't work.

Let's create the VAO!

At the top of your class, add the following:

```cs
private static uint _vao;
```

Then, in your `OnLoad` method, add:

```cs
_vao = _gl.GenVertexArray();
_gl.BindVertexArray(_vao);
```

What's going on here? First, we generate the VAO. Then, before we can use, or update it, we need to **bind** it.

> Binding is a term you will hear a lot with OpenGL, and in these tutorials too. Binding essentially means updating the current OpenGL state with the given object. So, in this example, when we bind the VAO, we update OpenGL's state so that the VAO we are currently using is now the VAO we have just created. If you don't bind an object before updating/using it, you will use or update the previously bound object instead, which can cause problems and headaches later.

Reviewer 3: ```suggestion
Binding is a term you will hear a lot with OpenGL, and in these tutorials too. Binding essentially means updating the current OpenGL state with the given object. So, in this example, when we bind the VAO, we update OpenGL's state so that the VAO in the state machine is now the VAO we have just created. If you don't bind an object before updating/using it, you will use or update the previously bound object instead, which can cause problems and headaches later.
```

I think "We are currently using" can be confusing

Author: I agree, although personally I prefer `the VAO currently stored in the state machine` as it just makes it a bit clearer that the state machine holds this specific VAO we have just set. But it's a minor change and I think the suggested change is still perfectly understandable. I'll commit your change and can I always change it later.

We're now ready to give the VAO some data. But first, we need a **vertex buffer**.

## Vertex Buffer Objects (VBOs)

> Here is the part where we actually give the GPU some data it can work with. For 95% of graphics applications, a vertex buffer is **required** to display anything on screen.

Reviewer 3: ```suggestion
Here is the part where we actually give the GPU some data it can work with. For most graphics applications, a vertex buffer is **required** to display anything on screen.
```

### What is a vertex buffer?
Before we can continue, it's handy to know what a vertex buffer is, and what it does.

Let's define both vertex, and buffer.

> * Vertex - A point where two lines, or edges meet.

Reviewer 3: (This is just a note, I'm not asking for this to be changed, I like it because it's simple to understand. I just want to comment this to make sure you are aware that the OpenGL concept of "vertex" is actually much more complex)

Note: A vertex in OpenGL isn't necessarily this. You can define the vertex attributes you want and use them however you wish to. For example, a vertex might have just a "texture coordinates" attribute, and then the vertex shader gets the position (to assign to `gl_Position`) from a texture. Another example is a small heightmap test project of mine, in which I had an empty VAO; No vertex attributes whatsoever! The vertex shader instead used `gl_VertexID` to organize the vertex into a flat XZ mesh pattern, and then the Y position came from reading the heightmap texture.

Author: Yeah, and this may be explored in later tutorials, I've been meaning to get into heightmaps and stuff myself too. For the purposes of this tutorial though, the vertex definition used here is helpful enough for the reader to understand.

* Buffer - A region in memory that can be accessed to be written to, or read from.

Knowing these definitions should give us a good idea of what a vertex buffer is. It's a region of memory, on the GPU, that stores points, which will get **rasterized** on-screen (usually in the form of triangles).

### Vertex data
Let's create the data that we will fill the vertex buffer with!

In your `OnLoad` function, just after you call `_gl.BindVertexArray()`, add the following:

```cs
float[] vertices =
{
     0.5f,  0.5f, 0.0f,
     0.5f, -0.5f, 0.0f,
    -0.5f, -0.5f, 0.0f,
    -0.5f,  0.5f, 0.0f
};
```

### What makes up a quad?

> In modern graphics programming, you are expected to use triangles. While you may see options for quads, or lines, these are mostly deprecated.

Reviewer 3: ```suggestion
In modern graphics programming, you are expected to use triangles, lines, or points. While you may see options for quads, these are obsolete.
```

In [the knronos docs for primitives](https://www.khronos.org/opengl/wiki/Primitive), quads are deprecated as OpenGL 3.0, and have been removed from 3.1 and above. Lines and points are still part of the core specification.

Therefore, a quad is made of two right-angle triangles. This can best be seen if we view the result in **wireframe** mode.

![Wireframe quad](../../../images/opengl/chapter1/wireframe-quad.png)

In the image, you can also see where the four vertices go in relation to the quad. While you won't *usually* be defining vertices yourself, it's still handy to know how it works.

### Creating the buffer
Now that we've got our vertex data, let's create the buffer!

At the top of your class, add the following:

```cs
private static uint _vbo;
```

Then, in your `OnLoad` method, under where the vertices are defined, add:

```
_vbo = _gl.GenBuffer();
_gl.BindBuffer(BufferTargetARB.ArrayBuffer, _vbo);
```

Much like the VAO, we need to first generate the buffer, and then bind it. Unlike vertex arrays, however, buffers need to be bound to a **target**. This allows you to bind buffers to different targets at the same time. Some of the more common targets include:

* `ArrayBuffer` - The vertex buffer target (which is what we're using here).

> * `ElementArrayBuffer` - The element array buffer target (which you'll see later).

Reviewer 3: If you wish to make this more specific, here's [the doc on binding targets](https://www.khronos.org/opengl/wiki/Buffer_Object#General_use).

In summary:
- `ArrayBuffer`: Note that `glVertexAttribPointer()` doesn't have any way to specify which VBO to target! That's because it targets what's currently bound to `ArrayBuffer`.
- `ElementArrayBuffer`: Uhhhh this one's a bit of a pain. Unlike the other buffer targets, this one is not stored as a "general OpenGL state", but rather is part of the currently bound VAO. `ElementArrayBuffer` controls "which buffer to use as an index buffer for the currently bound VAO".

Author: I personally think this is a bit too specific to put in this tutorial, however I could add a note to the shader attributes section with the `glVertexAttribPointer` information, as it seems useful to know.

* `UniformBuffer` - A uniform buffer, not used in this tutorial, but we'll get to it in a later tutorial.

In this case, we're binding to `ArrayBuffer` since we're creating a vertex buffer.

### Filling our buffer with data
Let's fill our buffer with some data! Before we do that though, you need to be aware of `unsafe` in C#.

> #### Unsafe C#

Reviewer 3: Good catch! Explaining what `unsafe` is and why it's necessary is important.

> Silk.NET heavily uses `unsafe` code. Don't worry, this won't make your computer explode, however it does exit out of the "safe" managed environment of C#, and enters a realm where undefined behaviour, segmentation faults, and strange results are more likely to occur if you are not careful.

Reviewer 3: > Don't worry, this won't make your computer explode

probably****

Reviewer: Yeah, this is a liability risk!

> Silk.NET heavily uses `unsafe` code. Don't worry, this won't make your computer explode, however it does exit out of the "safe" managed environment of C#, and enters a realm where undefined behaviour, segmentation faults, and strange results are more likely to occur if you are not careful.

Reviewer 3: ```suggestion
Silk.NET heavily uses `unsafe` code. Don't worry, this won't make your computer explode, however it does exit out of the "memory safe" managed environment of C#, and enters a realm where undefined behavior, segmentation faults, and strange results are more likely to occur if you are not careful. Since we're working with low-level APIs, and OpenGL is defined in plain C, some unsafe code will be necessary in order for us to be able to communicate with it from C#.
```

Reviewer 4: are we sure we want american english here? there's [precedent](https://github.com/dotnet/Silk.NET/blob/93f987aa40f99c3b6e347d3863fe2e524c2a9398/examples/CSharp/OpenGL%20Tutorials/Tutorial%203.2%20-%20Diffuse%20Lighting/Program.cs#L122) for both british and american english

Reviewer 3: I'm just gonna link [this commit](https://github.com/dotnet/Silk.NET/commit/4ae5a303e4238f3ed4fae199b6e64bb5ab2c0d0c)...

Reviewer 4: ah, so is american english the standard in Silk? if so we should really do another British ->American english pass (maybe we should write something into the bakery to convert binded function names and enum names to american english :^)

Reviewer 3: Wouldn't hurt to do so, but I don't think it's all that important either. We could always just ctrl-F "colour" through the entire project.

That commit still makes me chuckle lol

Reviewer: CAN WE PLEASE USE BRITISH ENGLISH PLEASE IT MAY NOT BE NORMAL IN OUR CODE BUT IT IS DEFINITELY THE PRECEDENT IN OUR LITERATURE SEE ALL OF THE BLOG POSTS USING BRITISH ENGLISH SURE THAT'S BECAUSE I WROTE THEM BUT THE POINT STILL STANDS BRITISH ENGLISH PLEASE THANK YOU THE ONLY REASON WE DON'T USE IT IN CODE IS BECAUSE WE'RE FOLLOWING MICROSOFT'S PRECEDENT THERE KTHX BYE

Author: happy to change to british english if desired

<?# Info "If you wish to use `Span` instead, and remain in `safe` mode, Silk.NET does support these too. However, I will be using `unsafe` in this tutorial instead, as this is both what I personally use, as well as what the samples use." /?>

> Unsafe mode is not enabled by default, so we need to enable it. To enable it, add `<AllowUnsafeBlocks>true</AllowUnsafeBlocks>` to the `PropertyGroup` in your project's `.csproj` file. You also **MUST** add `unsafe` to your `OnLoad` and `OnRender` methods, like so:

Reviewer 3: ```suggestion
Unsafe mode is not enabled by default, so we need to enable it. To enable it:
- If you're on Visual Studio 2022, open your project properties and under Build --> General, make sure the box that says "Unsafe code" is checked.
- You can also edit your project's `.csproj` file manually, by adding `<AllowUnsafeBlocks>true</AllowUnsafeBlocks>` to the `PropertyGroup`.

You also **MUST** add `unsafe` to your `OnLoad` and `OnRender` methods, like so:
```

Author: thanks for the VS suggestions, I have literally 0 idea how VS works so I just don't include it in the tutorials

```cs
public static unsafe void OnLoad() { ... }
```

If you forget to do either of these, you'll get compile errors. While our `OnRender` method does not contain any unsafe code yet, it will when we draw to the screen later, so it's best just to do it now so you don't forget to do it later.

Now let's fill the buffer!

> Add the following after you create the buffer:

Reviewer 3: ```suggestion
Add the following code to your `OnLoad` after you create the buffer:
```

Author: ```suggestion
Add the following to your `OnLoad` method:
```

I "standardized" this in my writing a while ago, I just forgot to change it earlier on. I will commit this and double check it's all standardized when I read through it and make some adjustments tomorrow.

```cs
fixed (float* buf = vertices)
    _gl.BufferData(BufferTargetARB.ArrayBuffer, (nuint) (vertices.Length * sizeof(float)), buf, BufferUsageARB.StaticDraw);
```

> Let's go over what's going on here. First, we `fix` the vertices data. This prevents the garbage collector from moving it around, so we can take a pointer to the data. We tell it the target we want, the `ArrayBuffer` in this case, give it the data length, the buffer pointer, and we choose `StaticDraw` for our usage.

Reviewer 3: ```suggestion
Let's go over what's going on here. First, we `fix` the vertex data. This prevents the garbage collector from moving it around, so we can take a pointer to the data. We then call `_gl.BufferData` and tell it the target we want, `ArrayBuffer` in this case, give it the data length (measured in bytes), the buffer pointer, and we choose `StaticDraw` for our usage hint.
```

Reviewer 4: Should we elaborate on what `StaticDraw` means? perhaps a list with small descriptions should be somewhere?

Author: Yeah, I'll add a small description of what static draw means, along with dynamic draw and stream draw.

I don't remember the others and I've never used them so I'll probably skip their explanation for this tutorial.

The funny thing is that most GL drivers don't even care what usage hint you give them, it's pretty much all the same now (from what I've heard).

> Now, run the program again, and if you get the same blue window, you've set the buffer data successfully!

Reviewer 3: (Not necessarily true, the blue window will remain whether you did it right or wrong)

Reviewer 4: > (Note: Not necessarily true, the blue window will remain whether you did it right or wrong)

The program may also segfault/throw an exception (along with the driver doing who knows what when a GL usage error occurs)

Reviewer 3: True that, but looking at the code it's waaaaay more likely that if you did something wrong it's gonna be something like forgetting to bind something, in which case it'll just silently fail.

I was just pointing it out though, I don't think it's wrong if that line remains as is

Author: The amount of times I have had driver crashes, segfaults, etc from failing to call `BufferData` correctly leads me to believe that if you do it wrong the program will crash.

You are right though that if the reader forgot to bind the buffer then they won't know, but that's why I include the "compare your code with the result" sections.

Reviewer 3: Really? I don't remember ever being annoyed by segfaults or driver crashes from calling `BufferData`. My most common mistake is probably getting the size wrong, by forgetting to multiply by sizeof(whatever), which wouldn't cause any issues other than leaving part of the buffer with garbage. I do see why it could be a common problem tho

Now, let's create the **element buffer object**.

## Element Buffer Objects (EBOs)

> Unlike a vertex buffer, an element buffer is not *strictly* required to display something on screen, however not using it will require making your vertex buffer a lot larger, with a lot more duplicate data.

Reviewer 3: ```suggestion
Unlike a vertex buffer, an element buffer is not *strictly* required to display something on screen, however not using it in some cases may require making your vertex buffer a lot larger, with a lot more duplicate data.
```

Just because I usually don't need to use EBOs lol

For those familliar with Direct3D terms, EBOs are more commonly referred to as **index buffers**, which is what this tutorial will refer to them from now on.

Take a look at the `vertices` array we defined earlier. You may have noticed that it only has four different points in it. Great! That's how you make a quad. But remember - that's not how this works. As mentioned earlier, a quad is made up of two triangles. But wait - that's six points... But we've only defined four. What's going on?

Well, that's where index buffers come in! An index buffer simply tells the GPU which points in the vertex buffer to use for each triangle.

Add the following to your `OnLoad` method:

```cs
uint[] indices =
{
    0u, 1u, 3u,
    1u, 2u, 3u
};

> ```

Reviewer 3: Any particular reason for using `uint` for indices? I usually use `ushort` or if that's not enough `uint`, given that those are _always_ 16-bit and 32-bit respectively

Reviewer 4: I think its best to suggest `ushort` by default, as its half the data

Reviewer 3: Okay yes, I'm gonna ask you to change it because afterwards you have to specify the element type, and you do so as `UnsignedInt`, therefore you need to make sure that your indices are actually 32-bit.

You can change it to `uint` if you wish, I'd still recommend `ushort`. You'd also have to update part of the "Drawing to the screen" section.

Author: I personally use uint for my indices which is why I used it here, but it makes little difference in practice. The main reason I use uint is to support more than 65k vertices, which some of the models I import have. Personally for the purposes of the tutorial I don't think it makes much difference but I'm happy to change it.

Reviewer 3: Ah sorry my bad; I got confused and somehow thought you were using native-sized integers (nuint). I'm a big dum dum. I'll mark this as resolved

_gl.BindBuffer(BufferTargetARB.ElementArrayBuffer, _ebo);
```

Then, we fill the buffer with data, much in the same way as we did with the vertex buffer.

Again, add the following to your `OnLoad` method:

```cs
fixed (uint* buf = indices)
    _gl.BufferData(BufferTargetARB.ElementArrayBuffer, (nuint) (indices.Length * sizeof(uint)), buf, BufferUsageARB.StaticDraw);
```

And that's it! In OpenGL, this is the common way you'll **create** buffers, from VBOs, to EBOs, to UBOs.

Run the program again, and if you still see the blue window, you've successfully created the EBO!

Now we can move onto the **shader**.

## Shaders
This tutorial won't go fully into detail about what shaders are and how they work, instead we'll leave that for a later tutorial, but in essence, a shader is a small program that runs on your GPU, that tells it how to process and display the data in our buffers.

There are many different types of shaders:
* Vertex
* Fragment (or Pixel, as used by Direct3D)
* Geometry
* Compute
* Tessellation

These shaders all serve different purposes, however the most commonly used shaders are the **vertex** and **fragment** shaders, and we will be using those in this tutorial.

> A vertex shader is run (invoked) for every vertex in the vertex buffer. The vertex buffer is where you perform transformations, such as world and camera matrices (which we'll get into in a later tutorial).
> 
> The fragment shader is invoked for every fragment of every vertex on screen. A fragment is essentially a pixel, and Direct3D even calls them pixel shaders. As you may expect, these shaders are a lot more intensive than vertex shaders, and can often be the cause of GPU slowdowns. Fragment shaders is where you perform the stuff that gets displayed on screen, such as texturing and many forms of lighting.

Reviewer 3: ```suggestion
A vertex shader is run (invoked) for every vertex in the vertex buffer. The vertex buffer is where you perform transformations such as translating, rotating or scaling and object. Once the vertex shaders are done running, our vertices are arranged into **primitives** (in our case, triangles), and the on-screen pixels these triangles occupy are filled by fragment shaders. This process is called **rasterization**.

The fragment shader is invoked for every fragment of every primitive on screen. A fragment is essentially a pixel, and Direct3D even calls them pixel shaders. As you may expect, these shaders are a lot more intensive than vertex shaders, and can often be the cause of GPU slowdowns, given that a single triangle may only invoke three vertex shaders, but can result in hundreds or thousands of fragment shader invocations!

Fragment shaders is where you perform the stuff that gets displayed on screen, such as texturing and many forms of lighting.
```

To keep the VS explanation simpler, and be more correct with the FS explanation

Author: yeah, this is much better.

> For the moment, this is all you need to know about shaders. You will learn more about shaders as we progress further in this tutorial, as knowing how shaders work and operate is a vital thing to know when it comes to modern graphics programming. The sky is the limit! (Or in our case, an endless cornflower blue void...)

Reviewer 3: > The sky is the limit! (Or in our case, an endless cornflower blue void...)

lol

Reviewer 3: _"Sometimes, I feel like my entire life is just one big static void..."_

### The shader code
Before we can create our shader objects, we need some shader code. OpenGL uses **GLSL** (Open**GL** **S**hading **L**anguage). Syntactically, it is quite similar to C. Don't worry though, there are no pointers in sight.

#### Creating the vertex shader

Add the following to your `OnLoad` method:

```cs
const string vertexCode = @"
#version 330 core

layout (location = 0) in vec3 aPosition;

void main()
{
    gl_Position = vec4(aPosition, 1.0);
}";
```

This is our vertex shader! Let's go through it line-by-line.

First, we must tell it which GLSL version we wish to use. Since OpenGL 3.3 (the version we are using), the GLSL version corresponds to the OpenGL version. We tell it we want `core`, as we are using Core OpenGL 3.3, rather than compatibility.

> Next, we define our **shader attributes**. In our example, we only have one, which is the position we defined in the vertex buffer. Remember - the vertex buffer is executed *for each* vertex in the vertex buffer, which is why we only need one value at a time. We define it at a manual "location", 0. While this is not necessary, we will be using it in this tutorial, as certain drivers (such as intel) like to use random locations if you do not specify one, which can cause many problems if you are not careful.

Reviewer 3: ```suggestion
Next, we define our **shader attributes**. In our example, we only have one, which is the position we defined in the vertex buffer. Remember - the vertex shader is executed *for each* vertex in the vertex buffer, which is why we only need one value at a time. We define it at a manual "location", 0. While this is not necessary, we will be using it in this tutorial, as certain drivers (such as intel) like to use random locations if you do not specify one, which can cause many problems if you are not careful.
```

> Much like C, we define our entry point, the `main` function. In it, we set the value of `gl_Position` to the vertex position. As we are not performing any transformations, we simply set the input value directly to the output value. `gl_Position`, which is a built-in shader variable, only accepts a `vec4` value though, so we must convert `aPosition` to `vec4` first. You may notice we use `1.0` instead of `0.0` though. While in this case it does not matter, it's a habit you should get into, as this will become very important when we start using matrices and transformations.

Reviewer 3: ```suggestion
Much like C, we define our entry point, the `main` function. In it, we set the value of `gl_Position` to the vertex position. As we are not performing any transformations, we simply set the input value directly to the output value. `gl_Position`, which is a built-in shader variable, only accepts a `vec4` value though, so we must convert `aPosition` to `vec4` first. You may notice we use `1.0` for the W value instead of `0.0` though. While in this case it does not matter, it's a habit you should get into, as this will become very important when we start using matrices and transformations.
```

#### Creating the fragment shader
Our fragment shader for this example is very simple.

Add the following to your `OnLoad` method:

```cs

> const string fragmentCode = @"
> #version 330 core
> 
> out vec4 out_color;
> 
> void main()
> {
>     out_color = vec4(1.0, 0.5, 0.2, 1.0);
> }";

Reviewer 3: ```suggestion
const string fragmentCode = @"
#version 330 core

out vec4 outColor;

void main()
{
    outColor = vec4(1.0, 0.5, 0.2, 1.0);
}";
```
To keep consistent naming with the variables in the vertex shader?

I usually name this one `FragColor`. Do as you please tho

Author: Yeah I have a strict naming style I stuck with long ago:

* Attributes - `aAttributeName`
* Uniforms - `uUniformName`
* Out variables - `shadername_variableName` (for example `frag_texCoord`)
* Fragment shader out variables - `out_variableName`

Not sure if there's some "recommended" naming scheme but imo this makes it very easy to understand what each variable does and is doing.

Reviewer 3: Ah, I usually name them like so:
- VS Input attributes: `vAttributeName`
- VS output or FS input attributes: `fAttributeName`
- FS output color attribute: `FragColor`
- Uniforms: just `UniformName`

I don't know any specific convention though. Do as you please.

> ```cs
> const string fragmentCode = @"
> #version 330 core
> 
> out vec4 out_color;
> 
> void main()
> {
>     out_color = vec4(1.0, 0.5, 0.2, 1.0);
> }";
> ```

Reviewer 3: ```suggestion
```cs
const string fragmentCode = @"
#version 330 core

out vec4 outColor;

void main()
{
    outColor = vec4(1.0, 0.5, 0.2, 1.0);
}";
```
```

To keep consistent naming with the variables in the vertex shader?

I usually name this one `FragColor`. Do as you please tho

#### Creating the shader objects
Let's create the shader objects!

Add the following to your `OnLoad` method:

```cs
uint vertexShader = _gl.CreateShader(ShaderType.VertexShader);
_gl.ShaderSource(vertexShader, vertexCode);
```

First, we create the shader object itself. Each shader object must be assigned a **type**, in this case the vertex shader. Then we simply give the shader the code we defined earlier.

Now we need to **compile** the shader.

Add the following to your `OnLoad` method:

```cs
_gl.CompileShader(vertexShader);

> _gl.GetShader(vertexShader, ShaderParameterName.CompileStatus, out int status);
> if (status != (int) GLEnum.True)

Reviewer 3: ```suggestion
_gl.GetShader(vertexShader, ShaderParameterName.CompileStatus, out int vStatus);
if (vStatus != (int) GLEnum.True)
```

to match the resulting code from 1.2-final-result

    throw new Exception("Vertex shader failed to compile: " + _gl.GetShaderInfoLog(vertexShader));
```

First, we compile the shader. Then, we need to check to make sure it has compiled correctly, and if it hasn't, it needs to let us know what our error is. `glGetShader` can give us a lot of different parameters, some of which we'll be using in later tutorials. For now, we are just interested in the compile status.

The compile status returns `1` if successful, and `0` if not. If it's not successful, we throw an exception. `glGetShaderInfoLog` gets the error string for the shader, and we output that in our exception. You don't *have* to throw an exception though, you can do whatever you wish with this result. Just don't try to continue further, as you will get a link error! (We'll get to this in a minute).

> We've now successfully created and compiled our vertex shader. To compile our fragment shader, simply copy and paste the code you just wrote. Just make sure to change the `ShaderType` to `FragmentShader`!
> 

Reviewer 3: ```suggestion
We've now successfully created and compiled our vertex shader. To compile our fragment shader, simply copy and paste the code you just wrote. Just make sure to change the `ShaderType` to `FragmentShader`!

```cs
uint fragmentShader = _gl.CreateShader(ShaderType.FragmentShader);
_gl.ShaderSource(fragmentShader, fragmentCode);
_gl.CompileShader(fragmentShader);

_gl.GetShader(fragmentShader, ShaderParameterName.CompileStatus, out int status);
if (status != (int) GLEnum.True)
    throw new Exception("Fragment shader failed to compile: " + _gl.GetShaderInfoLog(fragmentShader));
```

```

Author: I decided not to include the fragment shader code as it *should* be fairly obvious to the reader what needs to be copy-pasted and changed (as I even prompt them to do so and what to change), so I'd rather not include the duplicate code, unless you really think I should. If they get really stuck they could just look at the final result code.

Reviewer 3: I think you should include it, mostly for those that are just copy-pasting, but also to show where the `fragmentShader` variable comes from.

Author: Alrighty, I'll include it in the tutorial

#### Creating the program
Finally, we can create the **shader program**. In OpenGL, shaders are stored in a single **program** object.

First, add the following to the top of your file:

```cs
private static uint _program;
```

Then, add the following to your `OnLoad` method:

```cs
_program = _gl.CreateProgram();
```

Now that we've created the program, we now need to attach the shaders to it and **link** them to the program.

Add the following to your `OnLoad` method:

```cs
_gl.AttachShader(_program, vertexShader);
_gl.AttachShader(_program, fragmentShader);

_gl.LinkProgram(_program);

_gl.GetProgram(_program, ProgramPropertyARB.LinkStatus, out int lStatus);
if (lStatus != (int) GLEnum.True)
    throw new Exception("Program failed to link: " + _gl.GetProgramInfoLog(_program));
```

> Much like before, we must check to make sure that the program has linked correctly. Link errors are rare, most often caused by attempting to link with a shader that has failed to compile, which, assuming you implemented the checks earlier, should basically never happen.

Reviewer 3: ```suggestion
Much like before, we must check to make sure that the program has linked correctly. Link errors are most often caused by things like a typo in an `in` or `out` variable name, or attempting to link with a shader that has failed to compile, which, assuming you implemented the checks earlier, should basically never happen.
```

At least from my experience lol

Add the following to your `OnLoad` method:

```cs
_gl.DetachShader(_program, vertexShader);
_gl.DetachShader(_program, fragmentShader);
_gl.DeleteShader(vertexShader);
_gl.DeleteShader(fragmentShader);
```

Now, run your program again. If you get no errors, this has all worked successfully!

### Setting up the attributes
The last thing we need to do before we can begin drawing our quad to the screen is to set up the attributes. By default, OpenGL does not know how to send the data to our shader, so we must tell it.

<?# Warning "While fragment shaders (and other shaders) can have `in` attributes, the only ones you can directly set *outside* of a shader are the ones going into the vertex shader. The only way to set attributes in the fragment shader is to pass them through the vertex shader. Therefore, the attribute setup only affects the vertex shader and vertex buffer, not any other shader." /?>

Add the following to your `OnLoad` method:

```cs
_gl.UseProgram(_program);

const uint positionLoc = 0;
_gl.EnableVertexAttribArray(positionLoc);

> _gl.VertexAttribPointer(positionLoc, 3, VertexAttribPointerType.Float, false, 3 * sizeof(float), (void*) 0);

Reviewer 3: This code is related to the VAO. Readers need to understand that this code is configuring the VAO, and not the shader program.

This code is usually written together with the creation of the VAO.

Author: Hrm. I virtually always see this code right at the bottom of the vao setup, including in the official silk example. I didn't know that it was configuring specifically the VAO though, since I remember having issues with this if I didn't have a shader program bound, which made me assume it was configuring both.

(As a side note, personally, I absolutely despise VAOs and GL's attribute setup, I don't even use them outside of a single vao for the entire program. Give me input layouts any day of the week...)

Reviewer 3: Yup, `EnableVertexAttribArray` and `VertexAttribPointer` apply config changes to the current VAO. Nothing to do with the current shader program.

The only thing that connects the VAO to the shader program are the attribute locations. In this case, it's just a single attribute, at location 0, of type vec3.

(What?? You use a single VAO for your entire program?? If you have to go around enabling and disabling attributes and calling `VertexAttribPointer` whenever you need to change vertex type or source VBOs, that's very inefficient, and you should use multiple VAOs)

Author: I do yeah, probably could use multiple VAOs though, I'll have to experiment with it.

I've changed the explanation to make it clear that its updating the VAO specifically (and also removed the bit that tells you to set the shader program)

> The last thing we need to do before we can begin drawing our quad to the screen is to set up the attributes. By default, OpenGL does not know how to send the data to our shader, so we must tell it.

> This is done with the **attribute setup**. This sets various parameters in the VAO, which tells OpenGL how to read the vertex data in the vertex buffer. As vertex buffers can contain more than just position data (it can contain almost anything you want), it is vital that OpenGL knows how to separate out the individual bits of data, so it can pass it to the shader correctly.

Hopefully this is okay?

### Cleaning up
We're nearly there. The last thing we should do is to unbind the various buffers. While you don't *need* to do this, it will help to reduce the risk of changing a value for the wrong buffer.

Add the following code to your `OnLoad` method:

```
_gl.BindVertexArray(0);
_gl.BindBuffer(BufferTargetARB.ArrayBuffer, 0);
_gl.BindBuffer(BufferTargetARB.ElementArrayBuffer, 0);
```

Doing this means we've "un-bound" everything, so calling something like `BufferData` won't affect the buffers we've just created.

<?# Warning "You **MUST** unbind the vertex array first, before unbinding the other buffers. If you forget to do it in this order, the buffer will be unbound from the vertex array, meaning you'll see incorrect results when you render the object." /?>

If you want to see the resulting code so far, you can see it [here](../sources/1.2.7-finished-setup.html).

## Drawing to the screen
It's finally time to draw to the screen! That was a lot of setup work we just did there. Fortunately, that was the hard part. Drawing our result to the screen is now very easy.

Add the following to your `OnRender` method:

```cs
_gl.BindVertexArray(_vao);
_gl.UseProgram(_program);

> _gl.DrawElements(PrimitiveType.Triangles, 6, DrawElementsType.UnsignedInt, null);

Reviewer 3: ```suggestion
_gl.DrawElements(PrimitiveType.Triangles, 6, DrawElementsType.UnsignedInt, (void*) 0);
```

I think it's better if we avoid `null` on these numeric values. It is confusing enough that numeric values have to be passed as pointers, even though they don't point to anything! I hope for 3.0 we can turn all of these into `uint`s or something ;_;

Author: Yeah it would definitely be good to have it as a uint, or at least an overload for it. I often (including myself when I switched to silk) see people using `0` by itself and then being confused as to why it didn't work. While I personally use `null`, for the purposes of the tutorial I agree that `(void*) 0` is better. I'll leave this unresolved for now so I remember to update all the files containing this code (and not just the tutorial) when I make these changes tomorrow.

The last parameter is a pointer to the starting index of the indices. Since we want all the indices, we just set this value to `null`. (Equivalent to `(void*) 0`).

And that's it! Run your program and you should see a lovely orange rectangle on a blue background. Exciting, isn't it... Right...?

![Final result](../../../images/opengl/chapter1/final-result-t2.png)

While this may have seen like a lot of set up for a boring result, this code can render pretty much anything you want to the screen. It remains pretty much the same, whether you're rendering a basic quad like this, or a complex 3D model. All you need to change are the vertices & indices going in, and some more complex shader code to handle the transformations.

Once you get the basics understood and you can confidently use it, creating more complex stuff becomes a lot easier.

If you would like to view the full source code for the application, you can view it [here](../sources/1.2-final-result.html).

## Cleaning up

> When you're done with an OpenGL resource, you should delete it, to free GPU memory. For this though, it's not needed. The driver will automatically clean all created resources when your application closes. In fact, it's recommended that you *don't* manually delete resources when the application closes. The driver can free these objects a lot faster than manually removing them. If you have a lot of objects, manually deleting can cause the application to hang while the driver tries to free everything.

Reviewer 3: > The driver can free these objects a lot faster than manually removing them

Interesting, I did not know 🤔🤔

You **MUST** remember to remove unused objects while your application is running, however, as forgetting to do so will use more and more of the GPU's memory, potentially causing it to run out of memory. OpenGL isn't garbage collected! It requires manual memory management, a lot like C.

## Wrapping up
That was a lot to digest! You've hopefully learned a lot along the way though, and if there's bits you still don't understand, just go back and re-read the sections again to hopefully improve your understanding. Don't worry about it too much though, the learning curve is steep, and once you do understand it, you'll be writing graphics programs in no time!

Here's some stuff you can do now:

> * Move on to the [next tutorial](), where you'll learn how to create a GL context and display a quad on the screen.

Reviewer 3: Copy paste typo?

Author: Whoops, yes, I didn't even notice it when I read it just now, lol.
